### PR TITLE
Cache local reference file resolution

### PR DESCRIPTION
**Pull request by koxudaxi**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved local JSON Schema reference resolution for shared files, fragments, recursive schemas, absolute paths, and symlinked paths.
  * Prevented stale results when symlink targets change.
  * Safely handles failed resolutions and filesystem inspection errors.
  * Treats empty references consistently as root fragments.

* **Performance**
  * Added bounded caching to speed up repeated local reference resolution.

* **Tests**
  * Added regression, performance, and edge-case coverage for local reference handling.

### CODE REVIEW

File: src/datamodel_code_generator/reference.py [hunk 1]
```diff
@@ -583,6 +583,18 @@ def get_relative_path(base_path: PurePath, target_path: PurePath) -> PurePath:
...
 
 
+def _contains_symlink(path: Path) -> bool:
+    """Return whether resolving ``path`` can traverse an existing symlink."""
+    while path != path.parent:
+        try:
+            if path.is_symlink():
+                return True
+        except OSError:
+            return True
```
[the remark this answers is not in the document]
Comment (koxudaxi) [reply src/datamodel_code_generator/reference.py:594]: Fixed in 77f77264. `_contains_symlink()` now uses `Path.lstat()` with `stat.S_ISLNK()` and treats inspection errors as uncached. The regression test now raises from `lstat`; cache-capacity checks also short-circuit before filesystem inspection.

File: src/datamodel_code_generator/reference.py [hunk 2]
```diff
@@ -914,11 +929,24 @@ def resolve_ref(self, path: Sequence[str] | str) -> str:  # noqa: PLR0911, PLR09
             return f"{'/'.join(self.current_root)}#"
         if path_absolute_ref := self._resolve_path_absolute_local_ref(joined_path):
             joined_path = path_absolute_ref
-        if self.current_base_path and not self.base_url and joined_path[0] != "#" and not is_url(joined_path):
+        if (
+            (current_base_path := self.current_base_path)
+            and not self.base_url
+            and joined_path[0] != "#"
```
[the remark this answers is not in the document]
Comment (koxudaxi) [reply src/datamodel_code_generator/reference.py:935]: Fixed in a35ea663. Empty references now share the existing `#` early return, including `current_root` handling, with a regression test for both root contexts.

File: src/datamodel_code_generator/reference.py [hunk 3]
```diff
@@ -914,11 +929,24 @@ def resolve_ref(self, path: Sequence[str] | str) -> str:  # noqa: PLR0911, PLR09
...
-            joined_path = get_relative_path(self._base_path, resolved_file_path).as_posix()
+            cache_key = current_base_path, file_path
+            if (resolved_file_part := self._resolved_local_file_parts.get(cache_key)) is None:
+                local_file_path = Path(current_base_path, file_path)
+                resolved_file_path = local_file_path.resolve()
+                resolved_file_part = get_relative_path(self._base_path, resolved_file_path).as_posix()
+                if len(self._resolved_local_file_parts) < self._MAX_RESOLVED_LOCAL_FILE_PARTS and not _contains_symlink(
+                    local_file_path
+                ):
+                    self._resolved_local_file_parts[cache_key] = resolved_file_part
```
[the remark this answers is not in the document]
Comment (koxudaxi) [reply src/datamodel_code_generator/reference.py:948]: The cache is intentionally scoped to one parse cycle: each generation builds a new parser and `ModelResolver`, and watch regeneration uses a fresh resolver. Paths that already contain symlinks remain uncached so retargeting is followed. Converting an initially regular input tree to a symlink during an active parse is outside that lifecycle contract; detecting arbitrary ancestor mutations correctly would require filesystem validation on every cache hit and remove this fast path. I’m keeping the per-cycle snapshot semantics.
